docs/proposals/daemon-controller.md

```diff
@@ -0,0 +1,151 @@
+
+### Cluster-Level Applications
+Datastore: Users might want to implement a sharded datastore in their cluster. A few nodes in the cluster, labeled ‘datastore’, might be responsible for storing data shards, and pods running on these nodes might serve data. This architecture requires a way to bind pods to specific nodes, so it cannot be achieved using a Replication Controller. A daemon controller is a convenient way to implement such a datastore. At Google, D is an example of such an application.
+
+For other uses, see the related [feature request](https://github.com/GoogleCloudPlatform/kubernetes/issues/1518)
+
+## Functionality
+
+The Daemon Controller will support standard API features:
+- create
+  - The spec for daemon controllers will have a pod template field.  Note that daemon controllers will not have a nodename field.
+  - Using the pod’s node selector field, Daemon controllers can be restricted to operate over nodes that have a certain label. For example, suppose that in a cluster some nodes are labeled ‘database’. You can use a daemon controller to launch a datastore pod on exactly those nodes labeled ‘database’.
+  - The spec for pod templates that run with the Daemon Controller is the same as the spec for pod templates that run with the Replication Controller, except there will not be a ‘replicas’ field (exactly 1 daemon will be launched per node).
+  - We will not guarantee that daemon pods show up on nodes before regular pods - run ordering is out of scope for this controller
+  - The Daemon Controller will not guarantee that Daemon pods show up on nodes (for example because of resource limitations of the node), but will make a best effort to launch Daemon pods (like Replication Controllers do with pods)
+  - A daemon controller named “foo” will add a “controller: foo” annotation to all the pods that it creates
+  - YAML example:
+```YAML
+  apiVersion: v1
+  kind: DaemonController
+  metadata:
+    labels:
+      name: datastore
+    name: datastore
+  spec:
+    template:
+      metadata:
+        labels:
+          name: datastore-shard
+      spec:
+        node-selector: 
+          name: datastore-node
+        containers:
+          name: datastore-shard
+          image: kubernetes/sharded
+          ports:
+            - containerPort: 9042
+              name: main
+```
+  - commands that get info
+    - get (e.g. kubectl get dc)
+    - describe
+  - Modifiers
+    - delete
+    - stop: first we turn down the Daemon Controller foo, and then we turn down all pods matching the query “controller: foo”
+    - label
+    - update
+    - extension: rolling-update
```

Comment:
I wouldn't initially support rolling-update. Just make update work first. You could delete and re-recreate as discussed here:
https://github.com/GoogleCloudPlatform/kubernetes/issues/10202#issuecomment-114616708

It's worth describing the plan/idea for rolling update, however.

Daemon should contain a declarative representation of the desired state, which means that it should orchestrate rolling updates. See the discussion of how this would work for Deployment in #1743. If we added this functionality to node controller and never deleted daemon pods once placed (essentially custom forgiveness #1574), we'd never need to replace pods on existing nodes, so the one pod template in the Daemon's spec would be sufficient. We could detect deviating nodes that need to be updated by hashing the config and storing it in an annotation, as discussed in #10202 and #8683.

Rollout ordering is discussed below.